torch/overrides.py

```diff
@@ -1587,3 +1601,229 @@ def is_tensor_like(inp):
+        mode (:class:`TorchFunctionMode`): the mode to set as current mode
+        replace (:class:`TorchFunctionMode`): the mode to replace.  You can
+            use this argument to change the mode in a situation where you
+            know what the current mode is (and you are intentionally
+            overwriting it.)  If you don't know what the current mode is,
+            use ``ignore_preexisting`` instead.
+        ignore_preexisting (bool): if True, ignore any preexisting mode
+            and overwrite it with the passed mode.
+    """
+    if not (
+        mode is None or
+        isinstance(mode, TorchFunctionMode) or
+        (isinstance(mode, type) and issubclass(mode, torch.Tensor))
+    ):
+        raise TypeError(
+            "expected to get None, TorchFunctionMode or Tensor subclass as argument, got "
+            f"{type(mode)} instead"
+        )
+    old = _get_torch_function_mode()
+    # Short circuit.  It is valid to reset a mode to yourself, we won't error
+    # here
+    if old is mode:
+        yield
+        return
+    if old is not None and not ignore_preexisting and old is not replace:
+        if isinstance(mode, TorchFunctionMode):
+            help_text = (
+                'Use push_torch_function_mode instead.'
+            )
+        else:
+            help_text = (
+                'If you intended to completely override the preexisting mode, '
+                'pass ignore_preexisting=True.  This can result in unexpected '
+                'behavior; please consider making your mode compositional!'
+            )
+        raise ValueError(
+            'Attempted to enable_torch_function_mode, but there is already an '
+            f'active mode {old}.  {help_text}'
+        )
+    # NB: we don't require TorchFunctionMode since this is intended to also
+    # let you directly pass a Tensor subclass type to "mode-ify" it.
+    if not hasattr(mode, '__torch_function__'):
+        raise ValueError(
+            'The argument passed to enable_torch_function_mode must implement '
+            '__torch_function__'
+        )
+    _set_torch_function_mode(mode)
+    try:
+        yield
+    finally:
+        _set_torch_function_mode(old)
+
+
+@contextlib.contextmanager
+def push_torch_function_mode(ctor) -> Iterator[TorchFunctionMode]:
+    """
+    Context manager that pushes a :class:`TorchFunctionMode` onto the current
+    mode stack; see the class for more information on what modes are.  Stacked
+    modes can delegate to each other by invoking the ``__torch_function__``
+    method for the ``inner`` mode.
+
+    Args:
+        ctor: a function that when invoked as ``ctor(inner=...)`` produces
+            a :class:`TorchFunctionMode`.  If your :class:`TorchFunctionMode`
+            has no ``__init__`` implementation, you can simply pass the class
+            itself (e.g., ``push_torch_function_mode(MyMode)``); otherwise,
+            use ``partial`` to partially apply the constructor with all
```

Comment:
nit: partial -> functools.partial

To check my understanding, this is the approach to create a Mode that has some additional metadata attached?

Reply:
Yup